src/shogun/machine/FeatureDispatchCRTP.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Shubham Shukla
+ */
+
+#ifndef _FEATUREDISPATCHCRTP_H__
+#define _FEATUREDISPATCHCRTP_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/LinearMachine.h>
+
+namespace shogun
+{
+	class CFeatures;
+	class CLabels;
+#define IGNORE_IN_CLASSLIST
+
+	template <class P, class T>
+	IGNORE_IN_CLASSLIST class CDenseFeaturesDispatch : public T
+	{
+	public:
+		/** Default constructor */
+		CDenseFeaturesDispatch() : T()
+		{
+		}
+
+		virtual ~CDenseFeaturesDispatch()
+		{
+		}
+
+	protected:
+		virtual bool train_dense(CFeatures* data)
+		{
+			REQUIRE(
+			    data->get_num_vectors() == this->m_labels->get_num_labels(),
+			    "Number of training vectors (%d) does not match number of "
+			    "labels (%d)\n",
+			    data->get_num_vectors(), this->m_labels->get_num_labels())
+
+			// check for type of CFeatures, then call the appropriate template
+			// method
+			auto obj = this->template as<P>();
+			if (data->get_feature_type() == F_DREAL)
+				return obj->template train_machine_templated<float64_t>(
+				    data->as<CDenseFeatures<float64_t>>());
+			else if (data->get_feature_type() == F_SHORTREAL)
+				return obj->template train_machine_templated<float32_t>(
+				    data->as<CDenseFeatures<float32_t>>());
+			else if (data->get_feature_type() == F_LONGREAL)
+				return obj->template train_machine_templated<floatmax_t>(
+				    data->as<CDenseFeatures<floatmax_t>>());
+			else
+				SG_SERROR(
+				    "Feature-type (%d) must be of type F_SHORTREAL (%d), "
```

Comment:
could we introduce something that translates the feature type enum to a string and print the strings here instead?

E.g. `std::string feature_type(EFeatureType)` that returns the actual string, which we then can use to print such things in a human readable way?

Reply:
I would not print all options here, think if one type is added, then we need to change all error messages.
Just saw that "Training with dense features of provided type %s is not possible"